projects/chaos/src/relay_fsm.c

```diff
@@ -0,0 +1,154 @@
+#include "relay_fsm.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+
+#include "can.h"
+#include "can_ack.h"
+#include "can_msg_defs.h"
+#include "can_transmit.h"
+#include "chaos_events.h"
+#include "event_queue.h"
+#include "fsm.h"
+#include "misc.h"
+#include "relay_id.h"
+#include "status.h"
+
+#define RELAY_FSM_MAX_RETRIES 3
+
+typedef struct RelayFsmAckCtx {
+  RelayId id;
+  uint8_t retries;
+} RelayFsmAckCtx;
+
+typedef struct RelayFsmCtx {
+  RelayFsmAckCtx ack_ctx;
+  CANAckRequest request;
+} RelayFsmCtx;
+
+static RelayFsmCtx s_fsm_ctxs[NUM_RELAY_FSMS];
+
+static bool prv_guard(const FSM *fsm, const Event *e, void *context) {
+  RelayFsmCtx *fsm_ctx = context;
+  return e->data == fsm_ctx->ack_ctx.id;
+}
+
+static StatusCode prv_ack_open_callback(CANMessageID msg_id, uint16_t device, CANAckStatus status,
+                                        uint16_t num_remaining, void *context) {
+  RelayFsmAckCtx *ack_ctx = context;
+  if (status != CAN_ACK_STATUS_OK) {
+    if (ack_ctx->retries < RELAY_FSM_MAX_RETRIES) {
+      ack_ctx->retries++;
+      event_raise(CHAOS_EVENT_RETRY_RELAY, ack_ctx->id);
+      return STATUS_CODE_OK;
+    } else {
+      ack_ctx->retries = 0;
+      event_raise(CHAOS_EVENT_RELAY_ERROR, ack_ctx->id);
+      return STATUS_CODE_OK;
```

Comment:
Seems a little odd to have so many `return STATUS_CODE_OK`s. The difference between this function and `prv_ack_close_callback` seem limited. Is the only reason these are separate functions because they raise different events?

Reply:
Done. I switched to a single `return STATUS_CODE_OK` and modified into a single function.